### ISSUE TITLE
cxxopts set in the user's toolchain should be forwarded when using clang

### ISSUE THREAD

**Issue by udaya2899** [bazel-contrib/rules_cuda#344]
Related to #324 #331 

When using #331, although the toolchain is detected, the cxxopts set is not forwarded. i.e. I have the following in my `.bazelrc`:

```
build --cxxopt='-std=c++20'
build --host_cxxopt='-std=c++20'

# FYI, I use cuda with clang
build --@rules_cuda//cuda:compiler=clang
build --@rules_cuda//cuda:runtime=@local_cuda//:cuda_runtime_static
```

But the clang cuda compiler doesn't know about the cxxopt set to `std=c++20` and fails by not understanding what `std::remove_cvref` is which is a C++20 feature.

A workaround I did was to do:
```
build --@rules_cuda//cuda:copts='-std=c++20'
```

Ideally, the copts should be forwarded. Also, any reason we have copt instead of cxxopt? I'm asking because I want to set to c++20 only for C++, not for C.

There is a workaround mentioned above, but it has the problem of not being C++ specific.

**Comment by cloudhan** [comment]
Dup of #136

**Comment by mering** [comment]
@cloudhan I think this issue is different. When using `clang` as cuda compiler, doesn't it accept the same options as when using the same `clang` for c++?

**Comment by udaya2899** [comment]
Yes, let's keep this open to add support for this after #331 is merged?

**Comment by cloudhan** [comment]
In rules cuda clang toolchain is treated as a cuda toolchain, so they shall be treated as different, even tho clang is a valid toolchain for cpp and share the same flag set for cpp. For nvcc, is not a good idea to fake it as a clang toolchain with a wrapper and is non goal for this repo. Ideally, clang shall be treated different.

**Comment by udaya2899** [comment]
The request is to only do that for clang, not for nvcc. This can be part of or follow up to #331

**Comment by jsharpe** [comment]
> The request is to only do that for clang, not for nvcc. This can be part of or follow up to [#331](https://github.com/bazel-contrib/rules_cuda/pull/331)

This isn't a good idea as it fails the Liskov substitution principle; you should be able to swap out the compilers and not get differing behaviours due to propagation or not of flags.

**Comment by cloudhan** [comment]
I think the root cause of the difficulty is that there is no way to reliablely to query the flags specified from `--cxxopt`, since for our configuration, host flag and device flag are appended exactly the same as is. They are just missing from the `host_compile_flags` variable.

https://github.com/bazel-contrib/rules_cuda/blob/f7cbf72d1834ea4c7f2e56b9c472db4ee08f4902/cuda/private/toolchain_configs/clang.bzl#L219-L256

If we can query them, we can add a new variable `cc_compile_flags` and guard it with a feature `use_cxxopts`

**Comment by cloudhan** [comment]
And for nvcc, they will be prepended with `-Xcompiler` for host compiler. Then there will be no leaky design at all.

**Comment by mering** [comment]
> OK, problem solved, here we are: https://bazel.build/rules/lib/fragments/cpp#cxxopts

Yes, I think this is the way to go. We probably want this for `copts`, `cxxopts` and potentially `conlyopts` or `linkopts`.

Can you re-open the issue please to track this work?
